cpp/src/strings/search/contains_multiple.cu

```diff
@@ -0,0 +1,315 @@
+  if (d_strings.is_null(str_idx)) { return; }
+
+  // get the string for this tile
+  auto const d_str = d_strings.element<string_view>(str_idx);
+
+  auto const lane_idx    = idx % tile_size;
+  auto const tile_idx    = static_cast<cudf::thread_index_type>(threadIdx.x) / tile_size;
+  auto const num_targets = d_targets.size();
+
+  // size of shared_bools = num_targets * block_size
+  // each thread uses num_targets bools
+  extern __shared__ bool shared_bools[];
+  // bools for the current string
+  auto bools = working_memory == nullptr ? (shared_bools + (tile_idx * tile_size * num_targets))
+                                         : (working_memory + (str_idx * tile_size * num_targets));
+
+  // initialize result: set true if target is empty, false otherwise
+  for (auto target_idx = lane_idx; target_idx < num_targets; target_idx += tile_size) {
+    auto const d_target = d_targets.element<string_view>(target_idx);
+    if constexpr (tile_size == 1) {
+      d_results[target_idx][str_idx] = d_target.empty();
+    } else {
+      auto const begin = bools + (target_idx * tile_size);
+      thrust::uninitialized_fill(thrust::seq, begin, begin + tile_size, d_target.empty());
+    }
+  }
+  if constexpr (tile_size == cudf::detail::warp_size) { __syncwarp(); }
+
+  auto const last_ptr = d_first_bytes + unique_count;
+  for (size_type str_byte_idx = lane_idx; str_byte_idx < d_str.size_bytes();
+       str_byte_idx += tile_size) {
+    // search for byte in first_bytes array
+    auto const sptr     = d_str.data() + str_byte_idx;
+    auto const chr      = static_cast<u_char>(*sptr);
+    auto const byte_ptr = thrust::lower_bound(thrust::seq, d_first_bytes, last_ptr, chr);
+    // if not found, continue to next byte
+    if ((byte_ptr == last_ptr) || (*byte_ptr != chr)) { continue; }
+    // compute index of matched byte
+    auto const offset_idx = static_cast<size_type>(thrust::distance(d_first_bytes, byte_ptr));
+    auto map_idx          = d_offsets[offset_idx];
+    auto const last_idx = (offset_idx + 1) < unique_count ? d_offsets[offset_idx + 1] : num_targets;
+    // check for targets that begin with chr
+    while (map_idx < last_idx) {
+      auto const target_idx = d_indices[map_idx++];
+      auto const bool_idx   = (target_idx * tile_size) + lane_idx;
+      auto const found      = tile_size == 1 ? d_results[target_idx][str_idx] : bools[bool_idx];
+      if (!found) {  // not found before
+        auto const d_target = d_targets.element<string_view>(target_idx);
+        if ((d_str.size_bytes() - str_byte_idx) >= d_target.size_bytes()) {
+          // first char already checked, so just check the [1, end) chars match
+          auto const tp = d_target.data();
+          if (thrust::equal(thrust::seq, tp + 1, tp + d_target.size_bytes(), sptr + 1)) {
+            if constexpr (tile_size == 1) {
+              d_results[target_idx][str_idx] = true;
+            } else {
+              bools[bool_idx] = true;
```

Comment:
Line 155: `auto const bool_idx   = (target_idx * tile_size) + lane_idx;`
Line 166: `bools[bool_idx] = true`
Here the matched target can be any of the non-empty input targets, because the `d_indices` is for all the targets.

Line 128, 133, 134:
```cpp
128:   for (auto target_idx = lane_idx; target_idx < num_targets; target_idx += tile_size) {
133:      auto const begin = bools + (target_idx * tile_size);
134:      thrust::uninitialized_fill(thrust::seq, begin, begin + tile_size, d_target.empty());
```
Line 128 only initializes for a specific lane for specific targets. 
So line 156 will try to use uninitialized value.
```cpp
      auto const found      = tile_size == 1 ? d_results[target_idx][str_idx] : bools[bool_idx];
```

Example, there are two targets: ['a',  'b']
This lane only initialized for 'a' target
But when this lane try to search 'b' target, the cache for 'b' target is not initialized.

We need to add a `__syncwarp();` after the initialization.